[analyzer] Add std::variant checker

spaits opened the pull request:
As my BSc thesis I've implemented a checker for std::variant and std::any, and in the following weeks I'll upload a revised version of them here.

# Prelude

@Szelethus and I sent out an email with our initial plans here: https://discourse.llvm.org/t/analyzer-new-checker-for-std-any-as-a-bsc-thesis/65613/2 We also created a stub checker patch here: https://reviews.llvm.org/D142354.

Upon the recommendation of @haoNoQ , we explored an option where instead of writing a checker, we tried to improve on how the analyzer natively inlined the methods of std::variant and std::any. Our attempt is in this patch https://reviews.llvm.org/D145069, but in a nutshell, this is what happened: The analyzer was able to model much of what happened inside those classes, but our false positive suppression machinery erroneously suppressed it. After months of trying, we could not find a satisfying enhancement on the heuristic without introducing an allowlist/denylist of which functions to not suppress.

As a result (and partly on the encouragement of @Xazax-hun) I wrote a dedicated checker!

The advantage of the checker is that it is not dependent on the standard's implementation and won't put warnings in the standard library definitions. Also without the checker it would be difficult to create nice user-friendly warnings and NoteTags -- as per the standard's specification, the analysis is sinked by an exception, which we don't model well now.

# Design ideas

The working of the checker is straightforward: We find the creation of an std::variant instance, store the type of the variable we want to store in it, then save this type for the instance. When retrieving type from the instance we check what type we want to retrieve as, and compare it to the actual type. If the two don't march we emit an error.

Distinguishing variants by instance (e.g. MemRegion *) is not the most optimal way. Other checkers, like MallocChecker uses a symbol-to-trait map instead of region-to-trait. The upside of using symbols (which would be the value of a variant, not the variant itself itself) is that the analyzer would take care of modeling copies, moves, invalidation, etc, out of the box. The problem is that for compound types, the analyzer doesn't create a symbol as a result of a constructor call that is fit for this job. MallocChecker in contrast manipulates simple pointers.

My colleges and I considered the option of making adjustments directly to the memory model of the analyzer, but for the time being decided against it, and go with the bit more cumbersome, but immediately viable option of simply using MemRegions.

# Current state and review plan

This patch contains an already working checker that can find and report certain variant/any misuses, but still lands it in alpha. I plan to upload the rest of the checker in later patches.

The full checker is also able to "follow" the symbolic value held by the std::variant and updates the program state whenever we assign the value stored in the variant. I have also built a library that is meant to model union-like types similar to variant, hence some functions being a bit more multipurpose then is immediately needed.

I also intend to publish my std::any checker in a later commit.

spaits committed:
[analyzer] Add std::variant checker

Adding a checker that checks for bad std::variant type access.

NagyDonat reviewed (commented):
Very promising checker with straightforward and clean logic! I have lots of comments, but they are all minor/cosmetic issues.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/TaggedUnionModeling.h:
```diff
@@ -0,0 +1,128 @@
+
+  // Remove the information we know about the invalidate region.
+  // It is not relevant anymore.
+  State = std::accumulate(
+      Regions.begin(), Regions.end(), State,
+      [](ProgramStateRef State, const MemRegion *CurrentMemRegion) {
+        if (State->contains<TypeMap>(CurrentMemRegion)) {
+          State = State->remove<TypeMap>(CurrentMemRegion);
+        }
+        return State;
+      });
```
You don't need the `containts` there. You can just call the `remove` which is identity if the item is absent.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,312 @@
+        return getNthTemplateTypeArgFromVariant(
+            ArgType, TypeOut.getAsIntegral().getSExtValue());
+      default:
+        llvm_unreachable("An std::get's first template argument can only be a "
+                         "type or an integral");
```
`llvm_unreachable` should be only really rarely used. Only where it's mandated and justified.
Letting the compiler know that it could optimize assuming it's unreachable it a dangerous territory.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,312 @@
+      return false;
+    llvm::SmallString<128> Str;
+    llvm::raw_svector_ostream OS(Str);
+    OS << "std::variant " << ArgMemRegion->getDescriptiveName() << " held a(n) "
+       << StoredType->getAsString() << " not a(n) "
```
+1

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,312 @@
+
+const CXXConstructorDecl *
+getConstructorDeclarationForCall(const CallEvent &Call) {
+  const auto *ConstructorCall = dyn_cast<CXXConstructorCall>(&Call);
+  if (!ConstructorCall) {
+    return nullptr;
+  }
+  return ConstructorCall->getDecl();
```
```suggestion
  if (const auto *ConstructorCall = dyn_cast<CXXConstructorCall>(&Call))
    return ConstructorCall->getDecl();
  return nullptr;
```

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,312 @@
+  return ConstructorCall->getDecl();
+}
+
+bool isCopyConstructorCall(const CallEvent &Call) {
+  const CXXConstructorDecl *ConstructorDecl =
+      getConstructorDeclarationForCall(Call);
+  if (!ConstructorDecl) {
+    return false;
+  }
+  return ConstructorDecl->isCopyConstructor();
+}
```
```suggestion
bool isCopyConstructorCall(const CallEvent &Call) {
  if (const auto *CtorDecl = getConstructorDeclarationForCall(Call))
    return CtorDecl->isCopyConstructor();
  return false;
}
```

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,312 @@
+const TemplateArgument &getFirstTemplateArgument(const CallEvent &Call) {
+  const CallExpr *CE = cast<CallExpr>(Call.getOriginExpr());
+  const FunctionDecl *FD = CE->getDirectCallee();
+  assert(1 <= FD->getTemplateSpecializationArgs()->asArray().size() &&
+         "std::get should have at least 1 template argument!");
```
`getTemplateSpecializationArgs` might return null;
You don't need the `asArray` to have a `size()`.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,312 @@
+  return FD->getTemplateSpecializationArgs()->asArray()[0];
+}
+
+bool isStdType(const Type *Type, const std::string &TypeName) {
```
Are you sure you want to accept owning std::strings here?
A view wouldn't be enough?

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,312 @@
+    return false;
+  }
+
+  return (Decl->getNameAsString() == TypeName) && Decl->isInStdNamespace();
```
```suggestion
  return (Decl->getName() == TypeName) && Decl->isInStdNamespace();
```

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,312 @@
+static ArrayRef<TemplateArgument>
+getTemplateArgsFromVariant(const Type *VariantType) {
+  const auto *TempSpecType = VariantType->getAs<TemplateSpecializationType>();
+  assert(TempSpecType &&
+         "We are in a variant instance. It must be a template specialization!");
```
+1

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,312 @@
+
+static QualType getNthTemplateTypeArgFromVariant(const Type *varType,
+                                                 unsigned i) {
+  return getTemplateArgsFromVariant(varType)[i].getAsType();
```
Raw array accesses should be guarded by an assert right before.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,312 @@
+        return true;
+      }
+      if (Call.getNumArgs() != 1) {
+        return true;
```
I'm pretty sure it's not right.
We should bind something to the lvalue in evalCall.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,312 @@
+      }
+      SVal thisSVal = [&]() {
+        if (IsVariantConstructor) {
+          const auto *AsConstructorCall = dyn_cast<CXXConstructorCall>(&Call);
+          return AsConstructorCall->getCXXThisVal();
```
`dyn_cast` might return null, which is unconditionally dereferenced.
The same applies to the next case.

spaits replied:
In line 184 there is a variable `IsVariantConstructor`. It can only be true if the call is a constructor and it is a variant constructor (the `CallDescription` is needed to decide if it is for std::variant) the same is true for the next case but there std::variant's assignment operator is the one being checked for.

So in these cases `dyn_cast` should not return null.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,312 @@
+          const auto *AsMemberOpCall = dyn_cast<CXXMemberOperatorCall>(&Call);
+          return AsMemberOpCall->getCXXThisVal();
+        }
+        llvm_unreachable("We must have an assignment operator or constructor");
```
Just have an assert here instead of an unreachable.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,312 @@
+  }
+
+private:
+  // The default constructed std::variant must be handled separately
+  // by default the std::variant is going to hold a default constructed instance
+  // of the first type of the possible types
+  void handleDefaultConstructor(const CallEvent &Call,
+                                CheckerContext &C) const {
+
+    const auto *AsConstructorCall = dyn_cast<CXXConstructorCall>(&Call);
+    assert(AsConstructorCall && "A constructor call must be passed!");
+
+    SVal ThisSVal = AsConstructorCall->getCXXThisVal();
```
```suggestion
  // The default constructed std::variant must be handled separately
  // by default the std::variant is going to hold a default constructed instance
  // of the first type of the possible types
  void handleDefaultConstructor(const CXXConstructorCall &Call,
                                CheckerContext &C) const {
    SVal ThisSVal = AsConstructorCall->getCXXThisVal();
```

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,312 @@
+      }
+    }();
+
+    QualType RetrievedCanonicalType = RetrievedType.getCanonicalType();
+    QualType StoredCanonicalType = StoredType->getCanonicalType();
+    if (RetrievedCanonicalType == StoredCanonicalType) {
+      return true;
+    }
```
Shouldn't we peel off CV quals?
To me retrieving as `const T` is the same as `T` in this context for any `T`.

spaits replied:
If we have
`std::variant<int, char> v;std::get<const int>(v);`

steakhal replied:
Hm, I thought canonical means that it desugars the types aka. removes typedefs.
And a const qualified canonical is different from a non CV qualified canonical.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/TaggedUnionModeling.h:
```diff
@@ -0,0 +1,128 @@
+      if (OtherQTypeKnown) {
+        OtherQType = State->get<TypeMap>(ArgMemRegion);
+      } else {
+        return State->contains<TypeMap>(ThisRegion)
+                   ? State->remove<TypeMap>(ThisRegion)
+                   : State;
```
Yup, it should be identity if the item is not present.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/TaggedUnionModeling.h:
```diff
@@ -0,0 +1,128 @@
+    return State->set<TypeMap>(ThisRegion, WoPointer);
+  }();
+
+  if (State) {
+    C.addTransition(State);
+  } else {
+    C.addTransition(Call.getState()->remove<TypeMap>(ThisRegion));
+  }
```
Code like this is always fishy to me. The author should know if the State is null or not.
If we have a non-null State, it rarely turns into null afterwards, thus checks like this are usually codesmell.

steakhal reviewed (commented):
This is my first round of review. So far it looks good.
I didn't have much time, but that's all I had.

I think it would make sense to have tests for variants containing variants and playing with the active members. That would be interesting to see.

DonatNagyE commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,324 @@
+}
+
+bool isStdVariant(const Type *Type) {
+  return isStdType(Type, llvm::StringRef("variant"));
```
I don't think that you need to call the constructor `llvm::StringRef` explicitly. Also note that there is a subclass of `StringRef` called `llvm::StringLiteral` which calculates the length of a string literal in compile time (it's used as the type of `constexpr` variables) -- perhaps you'd want to use that.

spaits commented:
Thank you for your time and effort for editing my commit @Szelethus and for reviewing my commit @DonatNagyE @steakhal .

I have edited my commit based on your comments. These are the thing I have added or changed:

+ Added a function to decide 'a' or 'an' should be used in the reports.
+ Added test cases to verify that the
     an std::variant holds a std::variant is handled well `variantHoldingVariant`,
     a multiple active type changes are handled well `typeChangeThreeTimes`, `assignmentOperator`.
+ Replaced `std::string` to `StringRef` wherever I found that possible
+ Removed the IIFEs that made code readability worse.
+ Removed `assert`s and `llvm_unreachable` where it was used improperly to "test" the correctness of input.
+ Used llvm coding conventions (hopefully everywhere)
+ Using `dyn_cast_or_null` instead of checking for `nullptr` and then casting.
+ Spell checking and correcting sentences that were not clear or misguiding.

DonatNagyE commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,324 @@
+    std::string StoredTypeName = StoredType->getAsString();
+    std::string RetrievedTypeName = RetrievedType.getAsString();
+    OS << "std::variant " << ArgMemRegion->getDescriptiveName() << " held "
+       << indefiniteArticleBasedOnVowel(StoredTypeName[0]) << " \'"
+       << StoredTypeName << "\' not "
+       << indefiniteArticleBasedOnVowel(RetrievedTypeName[0]) << " \'"
```
~~Paranoia: can we have some crazy type (e.g. a lambda type) where the name is an empty string?~~

Does not matter, even if the type name is empty, we still have a `'\0'` that we can pass to the helper function.

DonatNagyE commented on clang/test/Analysis/std-variant-checker.cpp:
```diff
@@ -0,0 +1,349 @@
+  (void**)b;
+}
+
+void variantHoldingVariant() {
+  std::variant<std::variant<int, char>, std::variant<char, int>> v = std::variant<int,char>(25);
+  std::variant<int, char> v1 = std::get<std::variant<int,char>>(v);
+  std::variant<char, int> v2 = std::get<std::variant<char,int>>(v); // expected-warning {{std::variant 'v' held a 'std::variant<int, char>' not a 'class std::variant<char, int>'}}
+}
```
Beautiful testcase :1st_place_medal: 

Demonstrates the full glory of the type system...

DonatNagyE commented on clang/test/Analysis/std-variant-checker.cpp:
```diff
@@ -0,0 +1,349 @@
+//----------------------------------------------------------------------------//
+
+// Verifying that we are not invalidating the memory region of a variant if
+// a non inlined or inlined function takes it as a constant reference or po'int'er
```
Search&replace caught that poor unlucky `po'int'er`...

DonatNagyE commented on clang/test/Analysis/std-variant-checker.cpp:
```diff
@@ -0,0 +1,349 @@
+}
+
+// Verifying that we invalidate the mem region of std::variant when it is
+// passed as a non const reference or a po'int'er to a non inlined function.
```
Yet another `po'int'er`.

DonatNagyE commented on clang/test/Analysis/std-variant-checker.cpp:
```diff
@@ -0,0 +1,349 @@
+  std::variant<int, char> vtmp = v;
+  vtmp = 5;
+  int a = std::get<int> (vtmp);
+  (void*) a;
```
Why are you casting everything to `void*` (or even `void**` in one case)?

The standard idiom for "don't warn about this unused variable" is just `(void) variable`, which works for any type (unlike casts to `void*`, which would fail for e.g. a `double`).

DonatNagyE commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,312 @@
+      return false;
+    }
+
+    // Get the mem region of the argument std::variant and get what type
+    // information is known about it.
```
@spaits I think you forgot about this remark (but if you disagree with the suggestion, that's also fine).

DonatNagyE commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,324 @@
+    case TemplateArgument::ArgKind::Integral:
+      // In the natural number case we look up which type corresponds to the
+      // number.
+      if (std::optional<QualType> NthTemplate =
+              getNthTemplateTypeArgFromVariant(
+                  ArgType, TypeOut.getAsIntegral().getSExtValue()))
+        RetrievedType = *NthTemplate;
+      else
+        return false;
+      break;
```
```suggestion
      if (std::optional<QualType> NthTemplate =
              getNthTemplateTypeArgFromVariant(
                  ArgType, TypeOut.getAsIntegral().getSExtValue())) {
        RetrievedType = *NthTemplate;
        break;
      }
      [[fallthrough]];
```
What would you think about this alternative? (You don't have to use it if you prefer the original.)

DonatNagyE commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,324 @@
+    std::string RetrievedTypeName = RetrievedType.getAsString();
+    OS << "std::variant " << ArgMemRegion->getDescriptiveName() << " held "
+       << indefiniteArticleBasedOnVowel(StoredTypeName[0]) << " \'"
+       << StoredTypeName << "\' not "
```
```suggestion
       << StoredTypeName << "\', not "
```
Minor nitpick: my intuition suggests that there should be a comma before the "not" in this message. (Disclaimer: my intuition about comma use is fallible, especially in English.)

DonatNagyE commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,324 @@
+class StdVariantChecker : public Checker<eval::Call, check::RegionChanges> {
+  // Call descriptors to find relevant calls
+  CallDescription VariantConstructor{{"std", "variant", "variant"}};
+  CallDescription VariantAsOp{{"std", "variant", "operator="}};
```
This variable name is misleading, it it seems to mean "variant as op[erator]". Consider switching to e.g. `VariantAssignOp`.

DonatNagyE commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,324 @@
+        return true;
+      }
+
+      if (Call.getNumArgs() != 1)
+        return false;
```
Perhaps add a comment/FIXME to explain why you don't handle those constructors.

DonatNagyE commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,324 @@
+
+      SVal ThisSVal;
+      if (IsVariantConstructor) {
+        const auto *AsConstructorCall = dyn_cast<CXXConstructorCall>(&Call);
+        ThisSVal = AsConstructorCall->getCXXThisVal();
```
```suggestion
        const auto &AsConstructorCall = cast<CXXConstructorCall>(Call);
        ThisSVal = AsConstructorCall.getCXXThisVal();
```
(1) If you do not check the result of `dyn_cast` for nullness, then just use `cast` instead.
(2) `cast` is smart and can work on references instead of pointers (and btw `dyn_cast` also has an overload that takes a value and returns an `std::optional`).

DonatNagyE commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,324 @@
+        const auto *AsConstructorCall = dyn_cast<CXXConstructorCall>(&Call);
+        ThisSVal = AsConstructorCall->getCXXThisVal();
+      } else if (IsVariantAssignmentOperatorCall) {
+        const auto *AsMemberOpCall = dyn_cast<CXXMemberOperatorCall>(&Call);
+        ThisSVal = AsMemberOpCall->getCXXThisVal();
```
```suggestion
        const auto &AsMemberOpCall = cast<CXXMemberOperatorCall>(Call);
        ThisSVal = AsMemberOpCall.getCXXThisVal();
```

DonatNagyE commented on clang/lib/StaticAnalyzer/Checkers/TaggedUnionModeling.h:
```diff
@@ -0,0 +1,104 @@
+    if (!OtherQType) {
+      State = State->remove<TypeMap>(ThisRegion);
+    } else {
+      // When move semantics is used we can only know that the moved from
+      // object must be in a destructible state. Other usage of the object
+      // than destruction is undefined.
```
As a side project (in a separate commit) you could extend this checker with code that warns when a moved-from variant object is accessed.

DonatNagyE commented on clang/test/Analysis/std-variant-checker.cpp:
```diff
@@ -0,0 +1,349 @@
+using char_t = char;
+
+
+//----------------------------------------------------------------------------//
+// std::get
+//----------------------------------------------------------------------------//
```
Paranoia: add a case that tests `std::get()` used on an `std::pair()`.

NagyDonat reviewed (commented):
Yet another round of review, with even less significant remarks :)

spaits commented:
@steakhal what do you think about the current version of this commit?

steakhal commented:
@DonatNagyE I can see you all have a lot of PRs. I have limited time, but I think I can do two this week. Maybe one next week.
In what order should I go through the PRs?
I guess, this one should be in the lucky two.

DonatNagyE commented:
> @DonatNagyE I can see you all have a lot of PRs. I have limited time, but I think I can do two this week. Maybe one next week. In what order should I go through the PRs? I guess, this one should be in the lucky two.

The most urgent ones are https://github.com/llvm/llvm-project/pull/66207 (a "real" review) and https://github.com/llvm/llvm-project/pull/66647 (where nothing significant changed since you reviewed it last time). I hope that you can do those and this one.

Moreover, I'm planning to merge the commit https://github.com/llvm/llvm-project/pull/67572 because it's just an NFC cleanup and @gamesh411 already reviewed it.

spaits commented:
@steakhal we have also run the checker on multiple open source projects (a few weeks ago) and it did not crash. (It did not have any findings. There were only two C++ 17 projects and they had retrieved value from std::variant at most 20 times).

Could you please take another look at the commit and give feedback, when you'll have time for it?

steakhal reviewed (changes requested):
I had a quick look once again.
This this time I focused on the library boundaries and APIs.

spaits commented on clang/lib/StaticAnalyzer/Checkers/TaggedUnionModeling.h:
```diff
@@ -0,0 +1,104 @@
+  return State;
+}
+
+template <class TypeMap>
+void handleConstructorAndAssignment(const CallEvent &Call, CheckerContext &C,
+                                    const SVal &ThisSVal) {
```
There is going to be an std::any checker. There is shared code between the two. The value constructors and assignment operators are handled the same way. My plan for later is to develop a something frameworkish to handle all union like types.

spaits commented on clang/lib/StaticAnalyzer/Checkers/TaggedUnionModeling.h:
```diff
@@ -0,0 +1,104 @@
+namespace ento {
+namespace tagged_union_modeling {
+
+// The implementation of all these functions can be found in the file
+// StdVariantChecker.cpp under the same directory as this file.
+CallEventRef<> getCaller(const CallEvent &Call, CheckerContext &C);
+bool isCopyConstructorCall(const CallEvent &Call);
+bool isCopyAssignmentCall(const CallEvent &Call);
+bool isMoveAssignmentCall(const CallEvent &Call);
+bool isMoveConstructorCall(const CallEvent &Call);
+bool isStdType(const Type *Type, const std::string &TypeName);
+bool isStdVariant(const Type *Type);
+bool calledFromSystemHeader(const CallEvent &Call, CheckerContext &C);
```
This header has shared code for std::variant checker and an std::any checker which I will try to publish later.

steakhal replied:
Then why aren't these part of the CallEvent API?

spaits replied:
There are functions I did not move into CallEvent because they are called on a CallEvent, cast it to a subclass, then tell things about it. It would not be nice to have a function like this in the CallEvent super class. But if you think I should move them there then I can do that.

spaits commented on clang/lib/StaticAnalyzer/Checkers/TaggedUnionModeling.h:
```diff
@@ -0,0 +1,104 @@
+// corresponding custom data in the program state.
+template <class TypeMap>
+ProgramStateRef
+removeInformationStoredForDeadInstances(const CallEvent *Call,
```
This is called from the checkRegionChanges callback. That callback has a const CallEvent * parameter and that is passed here. I can modify my function to take a const CallEvent &, but it is only called from checkRegionChanges callbacks. Should I still modify it to reference?

steakhal replied:
I don't understand how some other API might affect this one.
I thought you could always transform a reference to a pointer and the other way around using the deref and addrof operators.

spaits replied:
Then I will modify my function accordingly. Should I take a look at the `checkRegionChanges` callback in general and consider modifying it to take `const CallEvent &` instead of `const CallEvent *` ? Do you think that would be a good task? I think I would have time for it and I would enjoy doing it.

Commit:
[NFC] Namespace, comment

Commit:
Move functions into CallEvent class

spaits committed:
Add extra line break for checker definiton

Co-authored-by: Balazs Benics <benicsbalazs@gmail.com>

spaits commented:
@steakhal thank you for the review. I have made the changes you have asked for.
 - Changed the name of header guard macro to reflect the path of the file
 - Used the new namespace opening syntax
 - Changed a function's argument, so it takes reference instead of pointer
   - It may would be a good task for me to see why the `checkRegionChanges` callback uses pointer instead of reference.
 - Moved a few functions into the class `CallEvent`
   - There are other functions I did not move into `CallEvent` because they are called on a `CallEvent`, cast it to a subclass then tell things about it. It would not be nice to have a function like this in the `CallEvent` super class. But if you think I should move them there then I can do that.

Commit:
Format changes

Commit:
Fix wrong token name

spaits commented on clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp:
```diff
@@ -0,0 +1,327 @@
+// Returns the CallEvent representing the caller of the function
+// It is needed because the CallEvent class does not contain enough information
+// to tell who called it. Checker context is needed.
+CallEventRef<> getCaller(const CallEvent &Call, const ProgramStateRef &State) {
```
I moved this function into the `CallEvent` class.

spaits commented:
gentle ping @steakhal

Commit:
Change a function name

spaits commented:
@steakhal I talked with @DonatNagyE and @dkrupp . We came to the conclusion, that the checker is only an alpha checker, would also not mean breaking changes to the analyzer and would not bring false positives. It also has been reviewed by @DonatNagyE . I would like to merge it in this form and later iterate on it. Would it be okay for you?

steakhal commented:
Yes, why not..
